File: composition.rb
class Address
    attr_reader :street, :city, :postcode

    def initialize(street, city, postcode)
        @street = street
        @city = city
        @postcode = postcode
    end

    def to_s
        return "#{@street}, #{@city}, #{@postcode}"
    end
end

class Person
    attr_reader :address

    def initialize(name, age, street, city, postcode)
        @name = name
        @age = age
        # Person "has-a" Address
        @address = Address.new(street, city, postcode)
        # @address = { street: street, city: city, postcode: postcode }
    end

    def to_s
        return "#{@name} is #{@age} years old, and lives at #{@address}"
    end

    # def full_address
    #     return @address.full
    # end
end


# addr = Address.new("12 Any Street", "Brisbane", 4000)
matt = Person.new("Matt", 49, "12 Any Street", "Brisbane", 4000)
# p matt
puts matt

File: composition.py
class Address:
    def __init__(self, street, city, postcode):
        self.street = street
        self.city = city
        self.postcode = postcode

    def __str__(self):
        return f"{self.street}, {self.city} {self.postcode}"


class Person:
    def __init__(self, name, age, street, city, postcode):
        self.name = name
        self.age = age
        # Person "has-a" Address, so we use composition rather than inheritance
        self.address = Address(street, city, postcode)

    def __str__(self):
        return f"{self.name} is {self.age} years old, and lives at {self.address}"


p1 = Person("Matt", 49, "12 Any Street", "Brisbane", 4000)
print(p1)
print(p1.address.postcode)
